esp/esdllib/esdl_script.cpp

```diff
@@ -2310,6 +2340,318 @@ class CEsdlTransformOperationFunction : public CEsdlTransformOperationWithChildr
+            break;
+        case Signed:
+            injectSigned(rawValue, txSummary);
+            break;
+        case Unsigned:
+            injectUnsigned(rawValue, txSummary);
+            break;
+        case Decimal:
+            injectDecimal(rawValue, txSummary);
+            break;
+        default:
+            break;
+        }
+    }
+    virtual bool exec(CriticalSection* crit, IInterface* preparedForAsync, IEsdlScriptContext* scriptContext, IXpathContext* targetContext, IXpathContext* sourceContext) override
+    {
+        IEspContext* espCtx = scriptContext->queryEspContext();
+        CTxSummary* txSummary = (espCtx ? espCtx->queryTxSummary() : nullptr);
+        if (!txSummary)
+            return true;
+        OptionalCriticalBlock block(crit);
+
+        if (m_value)
+        {
+            StringBuffer value;
+            sourceContext->evaluateAsString(m_value, value);
+            injectValue(value, txSummary);
+        }
+        return true;
+    }
+
+    virtual void toDBGLog () override
+    {
+    #if defined(_DEBUG)
+        DBGLOG(">%s> %s(%s)", m_traceName.str(), m_tagname.str(), m_info.m_name.str());
+    #endif
+    }
+};
+
+class CEsdlTransformOperationTxSummaryTimer : public CEsdlTransformOperationWithChildren
+{
+protected:
+    enum Mode { Append, Set, Accumulate };
+    TxSummaryCoreInfo m_info;
+    Mode              m_mode = Append;
+
+public:
+    CEsdlTransformOperationTxSummaryTimer(IXmlPullParser& xpp, StartTag& stag, const StringBuffer& prefix, IEsdlFunctionRegister* functionRegister)
+        : CEsdlTransformOperationWithChildren(xpp, stag, prefix, true, functionRegister, nullptr)
+        , m_info(stag, m_tagname, m_traceName, m_ignoreCodingErrors)
+    {
+        const char* mode = stag.getValue("mode");
+        if (!isEmptyString(mode))
+        {
+            if (strieq(mode, "set"))
+                m_mode = Set;
+            else if (strieq(mode, "accumulate"))
+                m_mode = Accumulate;
+            else if (!strieq(mode, "append"))
+                esdlOperationError(ESDL_SCRIPT_InvalidOperationAttr, m_tagname, VStringBuffer("invalid mode '%s'", mode), m_traceName, !m_ignoreCodingErrors);
+        }
+    }
+
+    virtual ~CEsdlTransformOperationTxSummaryTimer()
+    {
+    }
+
+    virtual bool exec(CriticalSection* crit, IInterface* preparedForAsync, IEsdlScriptContext* scriptContext, IXpathContext* targetContext, IXpathContext* sourceContext) override
+    {
+        IEspContext* espCtx = scriptContext->queryEspContext();
+        CTxSummary* txSummary = (espCtx ? espCtx->queryTxSummary() : nullptr);
+        if (!txSummary)
+            return true;
+        StringBuffer value;
+        OptionalCriticalBlock block(crit);
+        unsigned start = msTick();
+        processChildren(scriptContext, targetContext, sourceContext);
+        unsigned end = msTick();
+        unsigned delta = (end >= start ? end - start : UINT_MAX - start + end);
```

Comment:
I think there are JLIB classes that time things.

Reply:
Updated with CTimeMon